### PR TITLE
Common downloader

### PR DESCRIPTION
**Pull request by rabitt**
Just a proposal! If we go with this, I need some help with updating the tests @drubinstein @tkell 😬

### CODE REVIEW

**Comment by rabitt** [comment]
@drubinstein before I go any further - is this more or less what you had in mind?

File: mirdata/web_downloader.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+            A list of  RemoteFileMetadata tuples of uncompressed data.
+            If None, there is no uncompressed data to download.
+        info_message (str or None):
+            A string of info to print when this function is called.
+            If None, no string is printed.
+        force_overwrite (bool):
+            If True, existing files are overwritten by the downloaded files.
+
+    """
+    Path(save_dir).mkdir(exist_ok=True)
```
Comment (drubinstein) [inline mirdata/web_downloader.py]: Is there a chance that we will ever have zip, tar and file downloads all in the same dataset? If not, it'd probably be easier to split this up into separate functions depending on the download type and call those functions separately.

File: mirdata/web_downloader.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        tfile = tarfile.open(tar_path, 'r:gz')
+    else:
+        tfile = tarfile.TarFile(tar_path, 'r')
+    tfile.extractall(save_dir)
+    tfile.close()
+    if cleanup:
+        os.remove(tar_path)
+
+
+def force_delete_all(remote, data_home):
```
Comment (drubinstein) [inline mirdata/web_downloader.py]: I think this could stay in utils since it's not web specific.

File: mirdata/web_downloader.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        os.remove(zip_path)
+
+
+def download_tar_file(tar_remote, save_dir, force_overwrite, cleanup=False):
+    tar_download_path = download_from_remote(
+        tar_remote, save_dir, force_overwrite)
+    untar(tar_download_path, save_dir, cleanup=cleanup)
+
+
+def untar(tar_path, save_dir, cleanup=False):
```
Comment (drubinstein) [inline mirdata/web_downloader.py]: This could also stay in utils since it's not web download specific.

File: mirdata/web_downloader.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    return download_path
+
+
+def download_zip_file(zip_remote, save_dir, force_overwrite, cleanup=False):
+    zip_download_path = download_from_remote(
+        zip_remote, save_dir, force_overwrite)
+    unzip(zip_download_path, save_dir, cleanup=cleanup)
+
+
+def unzip(zip_path, save_dir, cleanup=False):
```
Comment (drubinstein) [inline mirdata/web_downloader.py]: same comment about leaving this in utils

**Comment by drubinstein** [comment]
@rabitt I left some comments with my review. This is more or less what I envisioned, just some nitpicking. Having a kitchen sink function is probably not going to scale well if we need to add other web downloaders. Grouping all the web download utilities in the same file is 💯 . I'm wondering if we should just name it downloader. Are you ever downloading from not the web nowadays?

File: mirdata/web_downloader.py (hunk 1 above)
Comment (rabitt) [reply mirdata/web_downloader.py]: seldom both zip and tar, but often zip + file and tar + file. 

For one simple reason, I'd like to keep one mega function - we're going to test everything in `web_downloader`, we're not going to test each `loader.download()`. If we can have each `loader.download()` be a very thin wrapper on top of `web_downloader.downloader()`, the lack of test coverage isn't as bad as if each `loader.download()` has it's own logic.

File: mirdata/web_downloader.py (hunk 3 above)
Comment (rabitt) [reply mirdata/web_downloader.py]: true. So far it's only used internally by other functions in web_download - do you still think it should move?

**Comment by rabitt** [comment]
@tkell @drubinstein - would either of you have time to help write tests for the new functions in `download_utils.py`? Not immediately urgent :)

**Comment by tkell** [comment]
@rabitt, yep, but probably not for a week or three!  Or David can do them, I am easy.

**Pull request merged**